staging/src/k8s.io/apimachinery/pkg/labels/selector.go

```diff
@@ -919,24 +919,34 @@ func SelectorFromSet(ls Set) Selector {
 	return SelectorFromValidatedSet(ls)
 }
 
+var (
+	// Create constants for hot-path selectors
+	value0Path, keyPath = func() (*field.Path, *field.Path) {
+		path := field.ToPath()
+		return path.Child("values").Index(0), path.Child("key")
+	}()
+)
+
 // ValidatedSelectorFromSet returns a Selector which will match exactly the given Set. A
 // nil and empty Sets are considered equivalent to Everything().
 // The Set is validated client-side, which allows to catch errors early.
 func ValidatedSelectorFromSet(ls Set) (Selector, error) {
 	if ls == nil || len(ls) == 0 {
 		return internalSelector{}, nil
 	}
-	requirements := make([]Requirement, 0, len(ls))
-	for label, value := range ls {
-		r, err := NewRequirement(label, selection.Equals, []string{value})
-		if err != nil {
-			return nil, err
+
+	var allErrs field.ErrorList
+	for k, v := range ls {
+		if err := validateLabelKey(k, keyPath); err != nil {
+			allErrs = append(allErrs, err)
+		}
+		if err := validateLabelValue(k, v, value0Path); err != nil {
+			allErrs = append(allErrs, err)
 		}
-		requirements = append(requirements, *r)
 	}
-	// sort to have deterministic string representation
-	sort.Sort(ByKey(requirements))
-	return internalSelector(requirements), nil
+
+	// TODO: validate labels are valid label values
+	return setSelector(ls), allErrs.ToAggregate()
```

Comment:
Great point. It seems like this may be desirable in some case (avoid copy), but not in other cases.

Do you think its worth making two constructors (one with copy, one without copy) and giving users the choice, or just always copying (some cost, but still much cheaper than current Selector logic)?

Reply:
Actually the copying seems worse than the current logic, surprisingly:

```
name \ time/op               /tmp/old     /tmp/nocopy  /tmp/copy
SelectorFromValidatedSet-48   315ns ± 0%   124ns ± 0%   423ns ± 0%
RequirementString-48         63.5ns ± 0%  63.8ns ± 0%  63.1ns ± 0%

name \ alloc/op              /tmp/old     /tmp/nocopy  /tmp/copy
SelectorFromValidatedSet-48    192B ± 0%      0B         336B ± 0%
RequirementString-48          24.0B ± 0%   24.0B ± 0%   24.0B ± 0%

name \ allocs/op             /tmp/old     /tmp/nocopy  /tmp/copy
SelectorFromValidatedSet-48    5.00 ± 0%    0.00         2.00 ± 0%
RequirementString-48           1.00 ± 0%    1.00 ± 0%    1.00 ± 0%
```

So it seems like we should either accept (and document) that the input Set is 'owned' by the selector once passed, create two methods for copy vs no copy, or do nothing and close the PR?

Reply:
Alternatively we could make Set just implement Selector directly. That would be a user facing API change though; there would be new public methods and users would need to change their code to not use Set.AsSelector() to get the benefits (AsSelector() basically would become DeepClone() at that point)